### PR TITLE
Java System.Logger Support

### PR DESCRIPTION
**Pull request by Trigtrig**
### Description

Linked issue: #221

This PR enables the use of Java's System.Logger together with tinylog.

### Definition of Done

- [x] I read [contributing.md](https://github.com/tinylog-org/tinylog/blob/v2.4/contributing.md)
- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including edge cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

#### There are no TODOS left in the code

I am not sure how to handle the module parameter in System.LoggerFinder#getLogger(String name, Module module). See also my comment in the TinylogLoggerFinder class. I would be happy to consider suggestions on this.

#### Maven build works including compiling, tests, and checks (`mvn verify`)

I have written some tests for dealing with null as parameter values.These are to check that my implementations of System.Logger#log(...) throw NPEs according to the documentation of System.Logger#log(...). Unfortunately, the spotbug configuration prevents this form of testing with the bug 'NP_NULL_PARAM_DEREF_ALL_TARGETS_DANGEROUS'. One possibility would be to include an exception for all test classes in the spotbug configuration. 
```xml
<Match>
	<!-- JUnit Tests -->
	<Source name="~.*Test\.java" />
	<Bug pattern="NP_NULL_PARAM_DEREF_ALL_TARGETS_DANGEROUS, DM_USELESS_THREAD, DP_DO_INSIDE_DO_PRIVILEGED, RV_RETURN_VALUE_IGNORED_BAD_PRACTICE, RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT, SIC_INNER_SHOULD_BE_STATIC_ANON, UPM_UNCALLED_PRIVATE_METHOD, URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD" />
</Match>
```
Or I remove the corresponding tests again. Your thoughts on this?

### Documentation

Additions or amendments for the [public documentation](https://github.com/tinylog-org/tinylog/wiki/Documentation):

```markdown

```

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/tinylog-org/tinylog/blob/v2.0/license.txt) (mandatory)
- [x] I agree that my GitHub user name will be published in the release notes (optional)

### CODE REVIEW

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+	@Override
+	public boolean isLoggable(final Level level) {
+		Objects.requireNonNull(level);
+
+		// Setting the log level of a message to OFF makes little sense and should result in the message not being logged.
+		if (level == Level.OFF) {
+			return false;
+		} else {
+			return LOGGING_PROVIDER.isEnabled(STACKTRACE_DEPTH, null, translate(level));
```
Comment (pmwmedia) [inline jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java:59]: Potential performance improvement: Store the minimum active severity level as `static final` like in [slf4j/TinylogLogger.java](https://github.com/tinylog-org/tinylog/blob/42c4b8e855dd765814ec2555f1b235ff34448e85/slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogLogger.java). This allows the JVM to identify which logging methods will never output anything at runtime and eliminate these calls completely from the byte code (see conclusion of https://tinylog.org/v2/benchmark/ for details).

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	@Override
+	public void log(final Level level, final Supplier<String> msgSupplier) {
+		Objects.requireNonNull(level);
+		Objects.requireNonNull(msgSupplier);
+
+		if (isLoggable(level)) {
+			LOGGING_PROVIDER.log(
+				STACKTRACE_DEPTH, null, translate(level), null, null, msgSupplier.get(), (Object[]) null
```
Comment (pmwmedia) [inline jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java]: You don't need to call the get() method. tinylog will do it for you.

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	private final ConcurrentMap<String, TinylogLogger> loggers = new ConcurrentHashMap<String, TinylogLogger>();
+
+	public TinylogLoggerFinder() {
+		super();
+	}
+
+	@Override
+	public System.Logger getLogger(final String name, final Module module) {
+		// How to handle modules as required by getLogger()?
+		// Should we check if there is an existing logger instance for name and module instead of just checking the name?
```
Comment (pmwmedia) [inline jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java]: I would just ignore the module. Log4j2 does it also ignore by the way.

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	private final String name;
+	private final Module module;
+
+	/**
+	 * @param name   the name of the logger
+	 * @param module the module for which the logger is being created
+	 */
+	public TinylogLogger(final String name, final Module module) {
+		this.name = Objects.requireNonNull(name);
+		this.module = Objects.requireNonNull(module);
```
Comment (pmwmedia) [inline jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java]: The module can be removed.

**Review by pmwmedia — CHANGES_REQUESTED** [review]

**Comment by pmwmedia** [comment]
Thank you very much for your great pull request. It looks pretty complete :)

> I am not sure how to handle the module parameter in System.LoggerFinder#getLogger(String name, Module module). See also my comment in the TinylogLoggerFinder class. I would be happy to consider suggestions on this.

You can just ignore the module. Other major logging frameworks do the same. See https://github.com/apache/logging-log4j2/blob/master/log4j-jpl/src/main/java/org/apache/logging/log4j/jpl/Log4jSystemLoggerFinder.java for example.

> I have written some tests for dealing with null as parameter values.These are to check that my implementations of System.Logger#log(...) throw NPEs according to the documentation of System.Logger#log(...). Unfortunately, the spotbug configuration prevents this form of testing with the bug 'NP_NULL_PARAM_DEREF_ALL_TARGETS_DANGEROUS'. One possibility would be to include an exception for all test classes in the spotbug configuration.

I fully agree with your suggestion. You can really add the SpotBugs exception to keep the tests as simple as possibly.

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+		TinylogLogger logger = loggers.get(name);
```
Comment (rbygrave) [inline jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java]: As an alternative this could be done via `loggers.computeIfAbsent( ... )`.  Is there a preference to do it this way over using `computeIfAbsent()` ?

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java (hunk 1 above)
Comment (Trigtrig) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java:59]: Ok, I looked at the example. I just don't see how the approach can be implemented in System.Logger. In System.Logger there are no different methods for TRACE to ERROR. There is basically only one log method, which gets the log level passed to it. So I know only at runtime whether the passed level can be logged.

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java (hunk 5 above)
Comment (Trigtrig) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java]: I would also like to use computeIfAbsent(...) but it does not work for Java 1.6. See the failed release build: `/home/runner/work/tinylog/tinylog/jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java:[32,69] method references are not supported in -source 1.6`

Comment (rbygrave) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java]: Oh yes, jdk 1.6.  Quite right.

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java (hunk 1 above)
Comment (pmwmedia) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java:59]: Sorry! You are right :) All other logging API adapters provide separate methods for each severity level. Somehow I have assumed that System.Logger has such methods, too.

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+		} else {
+			return LOGGING_PROVIDER.isEnabled(STACKTRACE_DEPTH, null, translate(level));
+		}
+	}
+
+	@Override
+	public void log(final Level level, final String msg) {
+		Objects.requireNonNull(level);
+
+		if (isLoggable(level)) {
```
Comment (pmwmedia) [inline jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java]: I have just noticed that we have a bug here. This call of `isLoggable()` adds an extra element to the stack trace and the stack trace depth of 2 won't be correct anymore. We need a stack trace of 3 inside of `isLoggable()` in this case. I propose either to replace `isLoggable()` by `level != Level.OFF && LOGGING_PROVIDER.isEnabled(STACKTRACE_DEPTH, null, translate(level))` or to add another internal `isLoggable()` method that passes a stack trace depth of 3 to the logging provider.

**Comment by pmwmedia** [comment]
Could you verify the email address that you used in your commits? It seems like you used an email address that is assigned to the GitHub account (https://github.com/settings/emails).

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java (hunk 6 above)
Comment (Trigtrig) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLogger.java]: Thanks, I will look at it again next weekend ;)

**Comment by Trigtrig** [comment]
> Could you verify the email address that you used in your commits? It seems like you used an email address that is assigned to the GitHub account (https://github.com/settings/emails).

I checked the email address again. I hope it is shown correctly now

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java (hunk 5 above)
Comment (rbygrave) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java]: Just a thought, this `jsl-tinylog` module depends on `System.Logger` and therefore has an implied dependency on JDK 9+ .  It seems that for this `jsl-tinylog` module we could use:
```
<maven.compiler.source>9</maven.compiler.source>
<maven.compiler.target>9</maven.compiler.target>
```
... which would enable the use of `computeIfAbsent()`?

**Comment by pmwmedia** [comment]
> I checked the email address again. I hope it is shown correctly now

Thank you! Now it is verified and matching with your GitHub account :)

File: jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java (hunk 5 above)
Comment (pmwmedia) [reply jsl-tinylog/src/main/java/org/tinylog/jsl/TinylogLoggerFinder.java]: I fully agree! Feel free to change it if you want :)

File: jsl-tinylog/pom.xml [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+    <packaging>bundle</packaging>
+
+    <parent>
+        <groupId>org.tinylog</groupId>
+        <artifactId>parent</artifactId>
+        <version>2.4-SNAPSHOT</version>
```
Comment (pmwmedia) [inline jsl-tinylog/pom.xml]: 2.5-SNAPSHOT (I just released 2.4 on Monday)

**Review by pmwmedia — CHANGES_REQUESTED** [review]
If you will change the version number (see my comment below), the build will work again :)

**Review by pmwmedia — APPROVED** [review]

**Comment by pmwmedia** [comment]
Thank you very much for your effort! I'm very happy about your great commits. I merge them with pleasure.
